File: Editor/Tools/SelectObjectTool.cs
using System;
using System.Threading.Tasks;
using McpUnity.Unity;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for selecting objects in the Unity Editor
    /// </summary>
    public class SelectObjectTool : McpToolBase
    {
        public SelectObjectTool()
        {
            Name = "select_object";
            Description = "Sets the selected object in the Unity editor by path or ID";
        }
        
        /// <summary>
        /// Execute the SelectObject tool with the provided parameters asynchronously
        /// </summary>
        /// <param name="parameters">Tool parameters as a JObject</param>
        public override Task<JObject> ExecuteAsync(JObject parameters)
        {
            // Extract parameters
            string objectPath = parameters["objectPath"]?.ToObject<string>();
            if (string.IsNullOrEmpty(objectPath))
            {
                return Task.FromResult(McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'objectPath' not provided", 
                    "validation_error"
                ));
            }
                
            // Log the execution
            Debug.Log($"[MCP Unity] Selecting object: {objectPath}");
            
            // Try to find the object by path in the hierarchy
            GameObject gameObject = GameObject.Find(objectPath);
            
            // If not found by path, try to find by name
            if (!gameObject)
            {
                // Find all game objects with the given name
                GameObject[] allObjects = UnityEngine.Object.FindObjectsByType<GameObject>(FindObjectsInactive.Include, FindObjectsSortMode.InstanceID);
                foreach (GameObject obj in allObjects)
                {
                    if (obj.name == objectPath)
                    {
                        gameObject = obj;
                        break;
                    }
                }
            }
            
            // If still not found, check if it's a numeric ID
            if (!gameObject && int.TryParse(objectPath, out int instanceID))
            {
                UnityEngine.Object obj = EditorUtility.InstanceIDToObject(instanceID);
                if (obj is GameObject)
                {
                    gameObject = obj as GameObject;
                }
            }
            
            // Select the object if found
            bool success = false;
            if (gameObject)
            {
                Selection.activeGameObject = gameObject;
                EditorGUIUtility.PingObject(gameObject);
                success = true;
            }
                
            // Create the response
            return Task.FromResult(new JObject
            {
                ["success"] = success,
                ["message"] = success 
                    ? $"Successfully selected object: {objectPath}" 
                    : $"Failed to find object: {objectPath}",
                ["type"] = "text"
            });
        }
    }
}

File: Server/src/tools/selectObjectTool.ts
import { z } from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { ToolDefinition } from './toolRegistry.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

export function createSelectObjectTool(mcpUnity: McpUnity, logger: Logger): ToolDefinition {
  const toolName = 'select_object';
  return {
    name: toolName,
    description: 'Sets the selected object in the Unity editor by path or ID',
    paramsSchema: z.object({
      objectPath: z.string().describe('The path or ID of the object to select (e.g. "Main Camera" or a Unity object ID)')
    }),
    handler: async ({ objectPath }): Promise<CallToolResult> => {
      const response = await mcpUnity.sendRequest({
        method: toolName,
        params: { objectPath }
      });
      
      if (!response.success) {
        throw new McpUnityError(
          ErrorType.TOOL_EXECUTION,
          response.message || `Failed to select object: ${objectPath}`
        );
      }
      
      return {
        success: true,
        message: response.message,
        content: [{
          type: response.type,
          text: response.message || `Successfully selected object: ${objectPath}` 
        }]
      };
    }
  };
}
